ir/src/constraints/mod.rs

```diff
@@ -0,0 +1,315 @@
+
+pub const MIN_CYCLE_LENGTH: usize = 2;
+
+// HELPER STRUCTS
+// ================================================================================================
+
+/// A struct containing the node index that is the root of the expression, the trace segment to
+/// which the expression is applied, and the constraint domain against which any constraint
+/// containing this expression must be applied.
+#[derive(Debug, Copy, Clone)]
+struct ExprDetails {
+    root_idx: NodeIndex,
+    trace_segment: TraceSegment,
+    domain: ConstraintDomain,
+}
+
+impl ExprDetails {
+    fn new(root_idx: NodeIndex, trace_segment: TraceSegment, domain: ConstraintDomain) -> Self {
+        Self {
+            root_idx,
+            trace_segment,
+            domain,
+        }
+    }
+
+    fn root_idx(&self) -> NodeIndex {
+        self.root_idx
+    }
+
+    fn trace_segment(&self) -> TraceSegment {
+        self.trace_segment
+    }
+
+    fn domain(&self) -> ConstraintDomain {
+        self.domain
+    }
+}
+
+// CONSTRAINTS
+// ================================================================================================
+
+/// Contains the graph representing all of the constraints and their subexpressions, the set of
+/// variables used in the integrity constraints, and a matrix for each constraint type (boundary,
+/// validity, transition), where each index contains a vector of the constraint roots for all the
+/// constraints of that type against the segment of the trace corresponding to that index. For
+/// example, transition constraints against the main execution trace, which is trace segment 0, will
+/// be specified by a vector in transition_constraints[0] containing a [ConstraintRoot] in the graph
+/// for each constraint against the main trace.
+#[derive(Default, Debug)]
+pub(super) struct Constraints {
+    /// Constraint roots for all boundary constraints against the execution trace, by trace segment,
+    /// where boundary constraints are any constraints that apply to either the first or the last
+    /// row of the trace.
+    boundary_constraints: Vec<Vec<ConstraintRoot>>,
+
+    /// Constraint roots for all validity constraints against the execution trace, by trace segment,
+    /// where validity constraints are any constraints that apply to every row.
+    validity_constraints: Vec<Vec<ConstraintRoot>>,
+
+    /// Constraint roots for all transition constraints against the execution trace, by trace
+    /// segment, where transition constraints are any constraints that apply to a frame of multiple
+    /// rows.
+    transition_constraints: Vec<Vec<ConstraintRoot>>,
+
+    /// A directed acyclic graph which represents all of the constraints and their subexpressions.
+    graph: AlgebraicGraph,
+
+    /// Variable roots for the variables used in integrity constraints. For each element in a
+    /// vector or a matrix, a new root is added with a key equal to the [VariableValue] of the
+    /// element.
+    variable_roots: VariableRoots,
```

Comment:
As far as I can tell, this is a helper map which is used only during the construction of the graph and is not really needed after the graph is constructed. If so, it may be good to find a better way to handle it in the future so that once constructed `Constraints` does not need to retain any extra info.

Reply:
Yes, this is true. I think there are still a few refactors that could be done, but I wanted to keep this PR as minimal as possible. Once the remaining issues related to this work are addressed this is worth revisiting (or possibly while addressing those)